Roslyakov Evgeny/oop.js

```diff
@@ -0,0 +1,266 @@
+
+/* Sizes and stuffing constants */
+var hamburgerOptions = {};
+
+hamburgerOptions.SIZE_SMALL = {
+    name: 'small size',
+    price: 50,
+    calories: 20
+};
+hamburgerOptions.SIZE_LARGE = {
+    name: 'large size',
+    price: 100,
+    calories: 40
+};
+hamburgerOptions.STUFFING_CHEESE = {
+    name: 'cheese',
+    price: 10,
+    calories: 20
+};
+hamburgerOptions.STUFFING_SALAD = {
+    name: 'salad',
+    price: 20,
+    calories: 5
+};
+hamburgerOptions.STUFFING_POTATO = {
+    name: 'potato',
+    price: 15,
+    calories: 10
+};
+
+/* Get the humburger size (name) */
+Hamburger.prototype.getSize = function () {
+    return this.getName.call(this);
+};
+
+/* Get the humburger stuffing */
+Hamburger.prototype.getStuffing = function () {
+    var stuffings = '';
+    this.stuffing.forEach(function (item) {
+        stuffings += item.name + ' ';
+    });
+    return stuffings.trim();
+};
+
+/* Calculate the humburger price */
+Hamburger.prototype.calculatePrice = function () {
+    var stufPrice = this.stuffing.reduce(function (acc, cur) {
+        return acc + cur.price;
+    }, 0);
+    this.price = this.size.price + stufPrice;
+    return this.price;
+};
+
+/* Calculate the humburger calories */
+Hamburger.prototype.calculateCalories = function () {
+    var stufCalories = this.stuffing.reduce(function (acc, cur) {
+        return acc + cur.calories;
+    }, 0);
+    this.calories = this.size.calories + stufCalories;
+    return this.calories;
+};
+
+
+//---------- Class Salad ----------//
+/*@constructor
+ * @param salad        the chosen salad constant
+ */
+function Salad(options) {
+    MenuItem.call(this, options);
+}
+
+/* Inheritance */
+Salad.prototype = Object.create(MenuItem.prototype);
+
+/* Salad constants */
+var saladOptions = {};
+
+saladOptions.CEASAR = {
+    name: 'Ceasar',
+    price: 100,
+    calories: 20
+};
+saladOptions.OLIVIER = {
+    name: 'Olivier',
+    price: 50,
+    calories: 80
+};
+
+
+//---------- Class Drink ----------//
+/*@constructor
+ * @param drink        the chosen drink constant
+ */
+function Drink(options) {
+    MenuItem.call(this, options);
+}
+
+/* Inheritance */
+Drink.prototype = Object.create(MenuItem.prototype);
+
+/* Drinks constants */
+var drinkOptions = {};
+
+drinkOptions.COKE = {
+    name: 'coke',
+    price: 50,
+    calories: 40
+};
+drinkOptions.COFFEE = {
+    name: 'coffee',
+    price: 80,
+    calories: 20
+};
+
+
+//---------- Class Order ----------//
+function Order() {
+    this.items = [];
+};
+
+/* Add new item(s) in order */
+Order.prototype.addItems = function() {
+    this.items = this.items.concat([].slice.call(arguments));
+    return this;
+};
+
+/* Delete selected item from the order */
+Order.prototype.rmItem = function(rmName) {
+    if(this.items.length !== 0) {
+        this.items = this.items.filter(function (item) {
+            if(!item.name) {
+                return item.size.name !== rmName;
```

Comment:
And the same, function Order doesn't have a property size. 
`var b = new Order();
b.addItems('a');
console.log(b.rmItem('a'))` 
will cause an error

Reply:
Functionality of the program allows only remove items by its name like 'Olivier' for salad or 'small size' for humburger. And we can add just items from our menu. So if you try to add and then remove the 'a' item it will cause error, but it is fine. You can't add to your order something not from the menu.